txnsync/incoming.go

```diff
@@ -0,0 +1,307 @@
+func (imq *incomingMessageQueue) clear(m incomingMessage) {
+	if m.peer != nil {
+		imq.enqueuedPeersMu.Lock()
+		defer imq.enqueuedPeersMu.Unlock()
+		delete(imq.enqueuedPeers, m.peer)
+	}
+}
+
+// incomingMessageHandler
+// note - this message is called by the network go-routine dispatch pool, and is not syncronized with the rest of the transaction syncronizer
+func (s *syncState) asyncIncomingMessageHandler(networkPeer interface{}, peer *Peer, message []byte, sequenceNumber uint64) (err error) {
+	// increase number of incoming messages metric.
+	txsyncIncomingMessagesTotal.Inc(nil)
+
+	// check the return value when we exit this function. if we fail, we increase the metric.
+	defer func() {
+		if err != nil {
+			// increase number of unprocessed incoming messages metric.
+			txsyncUnprocessedIncomingMessagesTotal.Inc(nil)
+		}
+	}()
+
+	incomingMessage := incomingMessage{networkPeer: networkPeer, sequenceNumber: sequenceNumber, encodedSize: len(message), peer: peer}
+	_, err = incomingMessage.message.UnmarshalMsg(message)
+	if err != nil {
+		// if we received a message that we cannot parse, disconnect.
+		s.log.Infof("received unparsable transaction sync message from peer. disconnecting from peer.")
+		return err
+	}
+
+	if incomingMessage.message.Version != txnBlockMessageVersion {
+		// we receive a message from a version that we don't support, disconnect.
+		s.log.Infof("received unsupported transaction sync message version from peer. disconnecting from peer.")
+		return errUnsupportedTransactionSyncMessageVersion
+	}
+
+	// if the peer sent us a bloom filter, decode it
+	if incomingMessage.message.TxnBloomFilter.BloomFilterType != 0 {
+		bloomFilter, err := decodeBloomFilter(incomingMessage.message.TxnBloomFilter)
+		if err != nil {
+			s.log.Infof("Invalid bloom filter received from peer : %v", err)
+			return errInvalidBloomFilter
+		}
+		incomingMessage.bloomFilter = bloomFilter
+		// increase number of decoded bloom filters.
+		txsyncDecodedBloomFiltersTotal.Inc(nil)
+	}
+
+	// if the peer sent us any transactions, decode these.
+	incomingMessage.transactionGroups, err = decodeTransactionGroups(incomingMessage.message.TransactionGroups, s.genesisID, s.genesisHash)
+	if err != nil {
+		s.log.Infof("failed to decode received transactions groups: %v\n", err)
+		return errDecodingReceivedTransactionGroupsFailed
+	}
+
+	if peer == nil {
+		// if we don't have a peer, then we need to enqueue this task to be handled by the main loop since we want to ensure that
+		// all the peer objects are created synchronously.
+		enqueued := s.incomingMessagesQ.enqueue(incomingMessage)
+		if !enqueued {
+			// if we can't enqueue that, return an error, which would disconnect the peer.
+			// ( we have to disconnect, since otherwise, we would have no way to syncronize the sequence number)
+			s.log.Infof("unable to enqueue incoming message from a peer without txsync allocated data; incoming messages queue is full. disconnecting from peer.")
```

Comment:
When it gets here (or the the error below for peer.incomingMessages.enqueue), could be because the system is under pressure. 
Why then spend valuable time to do `decodeTransactionGroups`?
`decodeTransactionGroups` is called, even before checking if the node has the capacity to process it.

Reply:
The system is a multithreading system.. which means that while we decode, the queue can be emptied out. I agree that we could have better load throttling, but I'm not sure if we can "predict" that we'll be failing here.